server/tests/ballot_comparison/test_cvrs.py

```diff
@@ -1744,7 +1780,148 @@ def test_hart_cvr_upload(
+
+    test_cases: List[TestCase] = [
+        {
+            "scanned_ballot_information_file_contents": HART_SCANNED_BALLOT_INFORMATION,
+            "expected_processing_status": ProcessingStatus.PROCESSED,
+            "expected_processing_error": None,
+            "expected_processing_work_progress": manifest_num_ballots,
+        },
+        {
+            "scanned_ballot_information_file_contents": HART_SCANNED_BALLOT_INFORMATION_MISSING_RECORDS,
+            "expected_processing_status": ProcessingStatus.PROCESSED,
+            "expected_processing_error": None,
+            "expected_processing_work_progress": manifest_num_ballots,
+        },
+        {
+            "scanned_ballot_information_file_contents": HART_SCANNED_BALLOT_INFORMATION_WITH_FORMAT_VERSION_ROW,
+            "expected_processing_status": ProcessingStatus.PROCESSED,
+            "expected_processing_error": None,
+            "expected_processing_work_progress": manifest_num_ballots,
+        },
+        {
+            "scanned_ballot_information_file_contents": "CvrId,UniqueId\n",
+            "expected_processing_status": ProcessingStatus.ERRORED,
+            "expected_processing_error": "Scanned ballot information CSV missing UniqueIdentifier column.",
+            "expected_processing_work_progress": 0,
+        },
+        {
+            "scanned_ballot_information_file_contents": "CvrGuid,UniqueIdentifier\n",
+            "expected_processing_status": ProcessingStatus.ERRORED,
+            "expected_processing_error": "Scanned ballot information CSV missing CvrId column.",
+            "expected_processing_work_progress": 0,
+        },
+        {
+            "scanned_ballot_information_file_contents": "#FormatVersion 1\nCvrId,UniqueId\n",
+            "expected_processing_status": ProcessingStatus.ERRORED,
+            "expected_processing_error": "Scanned ballot information CSV missing UniqueIdentifier column.",
+            "expected_processing_work_progress": 0,
+        },
+        {
+            "scanned_ballot_information_file_contents": "#FormatVersion 1\nCvrGuid,UniqueIdentifier\n",
+            "expected_processing_status": ProcessingStatus.ERRORED,
+            "expected_processing_error": "Scanned ballot information CSV missing CvrId column.",
+            "expected_processing_work_progress": 0,
+        },
+        {
+            "scanned_ballot_information_file_contents": "#FormatVersion 1\n",
+            "expected_processing_status": ProcessingStatus.ERRORED,
+            "expected_processing_error": "Please submit a valid CSV file with columns separated by commas.",
+            "expected_processing_work_progress": 0,
+        },
+        {
+            "scanned_ballot_information_file_contents": "",
+            "expected_processing_status": ProcessingStatus.ERRORED,
+            "expected_processing_error": "CSV cannot be empty.",
+            "expected_processing_work_progress": 0,
+        },
+    ]
+
+    for test_case in test_cases:
```

Comment:
I really like using table-driven tests as a way to consolidate redundant test logic! But I'd also ideally use some kind of subtest construct to better delineate the subtests and corresponding subtest errors, snapshot names, etc.

I decided to punt on setting up proper subtests to get this fix out in time for Washington's audits but can make a ticket to revisit

Reply:
I've never used subtests before in Python - would be curious to see what that looks like if you have any examples or docs.

Reply:
These docs cover the gist! https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests.

This note captures one of the key benefits for looped tests like the ones I've written:

> Without using a subtest, execution would stop after the first failure, and the error would be less easy to diagnose because the value of `i` wouldn’t be displayed

Reply:
Sweet, I wonder if that works with `pytest`

Reply:
Good q! There does appear to be a plugin for `pytest` subtest support: https://pypi.org/project/pytest-subtests. Opened an exploratory ticket: https://github.com/votingworks/arlo/issues/1560